src/MiniProfiler.Providers.StackExchange.Redis/RedisStorage.cs

```diff
@@ -0,0 +1,263 @@
+        /// <param name="orderBy">(Optional) The order to fetch profiler IDs in.</param>
+        public async Task<IEnumerable<Guid>> ListAsync(
+            int maxResults,
+            DateTime? start = null,
+            DateTime? finish = null,
+            ListResultsOrder orderBy = ListResultsOrder.Descending)
+        {
+            // TODO: start/finish not yet supported - will simply return all results
+
+            IEnumerable<RedisValue> ids;
+            switch (orderBy)
+            {
+                case ListResultsOrder.Ascending:
+                    ids = (await _database.ListRangeAsync(ProfilerResultListKey, maxResults + 1, -1).ConfigureAwait(false)).Reverse();
+                    break;
+                case ListResultsOrder.Descending:
+                default:
+                    ids = await _database.ListRangeAsync(ProfilerResultListKey, 0, maxResults - 1).ConfigureAwait(false);
+                    break;
+            }
+            return ids.Select(x => Guid.Parse(x));
+        }
+
+        /// <summary>
+        /// Asynchronously stores <paramref name="profiler"/> under its <see cref="MiniProfiler.Id"/>.
+        /// </summary>
+        /// <param name="profiler">The <see cref="MiniProfiler"/> to save.</param>
+        /// <remarks>
+        /// Should also ensure the profiler is stored as being un-viewed by its profiling <see cref="MiniProfiler.User"/>.
+        /// </remarks>
+        public async Task SaveAsync(MiniProfiler profiler)
+        {
+            var id = profiler.Id.ToString();
+            RedisKey key = ProfilerResultKeyPrefix.Append(id);
+            RedisValue value = profiler.ToJson();
+
+            await _database.StringSetAsync(key, value, expiry: CacheDuration).ConfigureAwait(false);
+
+            await _database.ListLeftPushAsync(ProfilerResultListKey, id).ConfigureAwait(false);
+            await _database.ListTrimAsync(ProfilerResultListKey, 0, ResultListMaxLength - 1).ConfigureAwait(false);
+            await _database.KeyExpireAsync(ProfilerResultListKey, CacheDuration).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Asynchronously returns a <see cref="MiniProfiler"/> from storage based on <paramref name="id"/>,
+        /// which should map to <see cref="MiniProfiler.Id"/>.
+        /// </summary>
+        /// <param name="id">The profiler ID to load.</param>
+        /// <returns>The loaded <see cref="MiniProfiler"/>.</returns>
+        /// <remarks>
+        /// Should also update that the resulting profiler has been marked as viewed by its
+        /// profiling <see cref="MiniProfiler.User"/>.
+        /// </remarks>
+        public async Task<MiniProfiler> LoadAsync(Guid id)
+        {
+            RedisKey key = ProfilerResultKeyPrefix.Append(id.ToString());
+            RedisValue value = await _database.StringGetAsync(key).ConfigureAwait(false); ;
+            return MiniProfiler.FromJson(value);
```

Comment:
I'm not so concerned with the dependency, but would hate for that layer to be *in the provider* - it would have to be some interface in the base .Shared library for serializing, but then you've got the problem of needing the user to have `n` library additions (e.g. another lib which *does* require protobuf) just for the serialization piece, so they're included `MiniProfiler.*`, `MiniProfiler.Providers.StackExchange.Redis`, and `<something>` which uses protobuf. IMO, that's too much, and pretty confusing.

Now we are changing up `StackExchange.Redis` to have a L1/L2 caching layer (currently in another library internally at Stack), so everyone can use it. I'd vote we can *remove* `protobuf` there if needed later, and match *that* serialization story. After all, the serialization target constraints are per-environment. For example most document stores are JSON, redis is bytes, relational is typed columns etc.

IMO, a `protobuf-net` dependency is one of the fastest, slimmest options and the simplest thing for a user to get setup.